### ISSUE TITLE
Add option to enable vSphere toplogy in rancher-vsphere RKE1/RKE2 charts

### ISSUE THREAD

**Issue by dkeightley** [rancher/rancher#40727]
**Is your feature request related to a problem? Please describe.**

Captured in https://github.com/rancher/rke2/issues/3468, the feature is enabled via flags which are not currently exposed as a helm value.

**Describe the solution you'd like**

Helm value to enable the topology plugin for the vSphere CSI

# Checklist for items that need completed

- [ ] Update Image-mirror repo
- [ ] Update manifests in rancher/vsphere-charts
- [ ] Verify Slack hook notified about chart "release" in relevant channels
- [ ] Update chart in rancher/charts
- [ ] Update chart in rancher/rke2-charts (RKE2 team)
- [ ] Updated RKE2 to consume new chart (RKE2 team)

**Comment by a-blender** [comment]
https://github.com/rancher/rancher/issues/40783 was closed because it's already being tracked in _this_ issue.

**Comment by a-blender** [comment]
Will resume work on this issue post rancher 2.7.x release and will bundle the vSphere topology update with the next round of image bumps.

**Comment by Sahota1225** [comment]
synched up @sergey and decided to move all vsphere issues to Q3.

**Comment by a-blender** [comment]
This has been moved to Q2 because it will be fixed as part of the latest CSI updates. When testing the chart on rke today, I will add a custom helm value for topology and ensure that deploys successfully.

**Comment by a-blender** [comment]
@dkeightley Could you enlighten me as to how you tested this? I asked here as well :) https://github.com/rancher/vsphere-charts/pull/42#pullrequestreview-1324406190 I'm testing this feature on an rke cluster and am not seeing the `TopologyKey` labels get set on the csinodes despite them being passed into the chart.

**Comment by a-blender** [comment]
## Test plan

Test steps
* Run rancher instance of `v2.7-head` on a vsphere node
* Provision a 1 cp/etcd,2 worker rke vsphere node driver cluster via the UI.
* Install the CSI chart with the topology plugin and test values for k8s region and zone. Verify the chart yaml and labels set on the csinodes to ensure the plugin is functioning properly

test file
```
vCenter:
  host: "<host>"
  port: <port>
  insecureFlag: true
  clusterId: "ablender-test-cluster-id"
  datacenters: "<datacenter>"
  username: "<username>"
  password: "<password>"

# labels for regions and zones
labels:
  region: k8s-region-test
  zone: k8s-zone-test

topology:
  enabled: true
```

Installed CSI
```
helm install -f values.yaml --kubeconfig ~/Downloads/ablender-126.yaml csi ./charts/rancher-vsphere-csi
```

Chart deployed successfully and pods are healthy, and based on [the VMware topology docs](https://docs.vmware.com/en/VMware-vSphere-Container-Storage-Plug-in/3.0/vmware-vsphere-csp-getting-started/GUID-162E7582-723B-4A0F-A937-3ACE82EAFD31.html), I verified that topology was `enabled:true` in the chart yaml.

However, cis nodes did _not_ have the `TopologyKeys` labels on the cisnodes.
```
kubectl get csinodes -o jsonpath='{range .items[*]}{.metadata.name} {.spec}{"\n"}{end}' --kubeconfig ~/Downloads/ablender-126.yaml
ablender-126-cp-etcd1 {"drivers":[{"allocatable":{"count":59},"name":"csi.vsphere.vmware.com","nodeID":"4208d0e8-3e5f-214b-054c-b4c65f5e7cba","topologyKeys":null}]}
ablender-126-worker1 {"drivers":[{"allocatable":{"count":59},"name":"csi.vsphere.vmware.com","nodeID":"4208d41a-b9d7-fb66-4499-26831ac22cb1","topologyKeys":null}]}
ablender-126-worker2 {"drivers":[{"allocatable":{"count":59},"name":"csi.vsphere.vmware.com","nodeID":"42081a1c-9385-20ab-1fd6-03368b219d78","topologyKeys":null}]}
```

The values are being passed to the chart but the `TopologyKeys` label is not getting set on the nodes somehow. Could you enlighten me as to how you tested this? I asked here as well :) https://github.com/rancher/vsphere-charts/pull/42#pullrequestreview-1324406190

**Comment by a-blender** [comment]
Will poke around to see if I can create categories/tags in the vsphere backend infra so QA can test functionality.

**Comment by brandond** [comment]
I believe it is on the end user to set the topology keys themselves by creating tags on the vSphere datacenter and server resources, as per the upstream docs. The vsphere configuration file for the Kubernetes cluster must then list the tags that hold the region and zone values, in the `topology-categories` value of the `Labels` section. I'm not sure if any of these tags are set in our vSphere environment?

![image](https://github.com/rancher/rancher/assets/370103/2bba48cb-1304-44f3-9860-be770de7ede4)

**Comment by a-blender** [comment]
@brandond I've setup the categories/tags on our backend vsphere infra and followed the instructions from the [vmWare docs](https://docs.vmware.com/en/VMware-vSphere-Container-Storage-Plug-in/3.0/vmware-vsphere-csp-getting-started/GUID-162E7582-723B-4A0F-A937-3ACE82EAFD31.html) to set zoning on the downstream cluster nodes. I haven't been able to see `TopologyKeys` get set on the nodes so far, so I'm assuming it's not working? vmWare also doesn't make it clear whether these labels are required for functionality of the plugin.

vsphere categories
<img width="1756" alt="image" src="https://github.com/rancher/rancher/assets/12955547/dc3db660-63fc-4707-af43-eb34bb884de8">

vsphere Tags
-coming-

I reprovisioned my rke cluster with a node template that included a tag for k8s region-1 / zone-A
<img width="1118" alt="image" src="https://github.com/rancher/rancher/assets/12955547/50b09145-5e8e-46cf-b8c5-847b23b47151">

<img width="1118" alt="image" src="https://github.com/rancher/rancher/assets/12955547/5c9c698f-5d89-4415-8766-90b3e65b9545">

I installed CSI again
```
helm install -f values.yaml --kubeconfig ~/Downloads/ablender-126.yaml csi ./charts/rancher-vsphere-csi
```

No `TopologyKeys` were set like before
```
ablender-126-cp-etcd1 {"drivers":[{"allocatable":{"count":59},"name":"csi.vsphere.vmware.com","nodeID":"4208d0e8-3e5f-214b-054c-b4c65f5e7cba","topologyKeys":null}]}
ablender-126-worker1 {"drivers":[{"allocatable":{"count":59},"name":"csi.vsphere.vmware.com","nodeID":"4208d41a-b9d7-fb66-4499-26831ac22cb1","topologyKeys":null}]}
ablender-126-worker2 {"drivers":[{"allocatable":{"count":59},"name":"csi.vsphere.vmware.com","nodeID":"42081a1c-9385-20ab-1fd6-03368b219d78","topologyKeys":null}]}
```

**Comment by brandond** [comment]
The tags need to be placed on the datacenter and cluster in vsphere, not on the VM nodes.

**Comment by a-blender** [comment]
@brandond So does this feature only work for imported clusters? I have VMs setup in vsphere but no clusters are listed so I can't tag them. Do you mean tag the host or cp node?

**Comment by a-blender** [comment]
I investigated, and discovered we currently only have a Datacenter and EXSi hosts setup which Rancher can deploy cluster nodes onto. We don't have any clusters setup in vsphere to tag. We'll want to get that infra setup at some point, but for now we can just verify the chart deploys without error.

@thaneunsoo Please test this using [test plan](https://github.com/rancher/rancher/issues/40727#issuecomment-1558138676) and verify the CSI chart deploys and runs successfully with topology enabled. The `TopologyKeys` on csinodes will be null for now until the infra is setup.

**Comment by bsctl** [comment]
@a-blender 

> Test steps
> 1. Run rancher instance of v2.7-head on a vsphere node
> 2. Provision a 1 cp/etcd,2 worker rke vsphere node driver cluster via the UI.
> 3. Install the CSI chart with the topology plugin and test values for k8s region and zone. Verify the chart yaml and labels set on the csinodes to ensure the plugin is functioning properly.

Interested into this feature, having a multi DC vSphere setup currently running Rancher Prime 2.7.3 where this feature is not yet available. 

Trying to follow the steps above:

In step 2, creating the downstream cluster, we have the option to configure vSphere or Default or External cloud provider. If we leave the Default, the default cloud provider is used but step3 fails because the port used by vSphere provider is already used by the default rancher provider. Any suggestion?

**Comment by a-blender** [comment]
@bsctl Please reach out to @jiaqiluo with Rancher Prime questions or create a new GH issue. Thanks!

**Comment by bsctl** [comment]
@jiaqiluo RKE2

**Comment by brandond** [comment]
If using RKE2, you should chose the vsphere cloud provider when provisioning the cluster. The May releases will include topology support. If you use the default but then try to install the chart later from Rancher Apps, you will run into port conflicts due to running multiple cloud provider implementations.

**Comment by bsctl** [comment]
@brandond yep that’s my issue. So the resolution is to update RKE2? Please advice. Thanks

**Comment by brandond** [comment]
Yes, once you update to the may release, you should be able to configure the vsphere charts bundled with rke2 to enable topology support. Note that the releases are currently in QA and not yet available.

**Comment by bsctl** [comment]
@brandond any chance to get it working by using current RKE2 version? 

`v1.25.9+rke2r1", GitCommit:"a1a87a0a2bcd605820920c6b0e618a8ab7d117d4", GitTreeState:"clean", BuildDate:"2023-04-12T19:05:26Z"`

And manually hack the deployment as described in official vSphere [doc](https://docs.vmware.com/en/VMware-vSphere-Container-Storage-Plug-in/3.0/vmware-vsphere-csp-getting-started/GUID-162E7582-723B-4A0F-A937-3ACE82EAFD31.html#GUID-73D106A3-1D8A-4CDC-9762-6CB35A65B0B4)?  We're tring to do it but withous success, so just asking to avoid wasting time. Thanks.

**Comment by brandond** [comment]
There is no way to enable it in the version of the chart currently bundled with RKE2, which means you can't enable it at cluster creation time. You'd need to disable the rancher-vsphere-csi chart and deploy the newer version manually.

**Comment by bsctl** [comment]
@brandond how to disable the cloud provider at creation time? I do not want to use UI, I'm using the terraform module

```
resource "rancher2_cluster_v2" "main" {
  name                  = var.cluster_name
  kubernetes_version    = var.rke2_version
  enable_network_policy = false

  default_pod_security_policy_template_name = ""

  rke_config {
    chart_values = <<-EOT
      rke2-calico:
        installation:
          calicoNetwork:
            bgp: Enabled
            ipPools:
              - blockSize: 26
                cidr: 172.12.0.0/16
                encapsulation: IPIPCrossSubnet
      EOT
  }
}
```

with the config above, the RKE2 cluster is provisioned with Default RKE2 Embedded cloud provider. I do not want enable any cloud provider at creation time, so I can install the vsphere provider by my own.

**Comment by brandond** [comment]
I'm not sure how to do that from the teraform provider. You would need to disable the built-in cloud provider, and then specify additional manifests to deploy the rancher-vsphere-cpi and rancher-vsphere-csi charts. It's probably going to be a fair bit of work, when you could just wait a couple days for the new releases to be out.
